File: 03-march/22-determine-whether-matrix-can-be-obtained-by-rotation/solution.py
class Solution:
    def findRotation(self, mat: list[list[int]], target: list[list[int]]) -> bool:
        def rotate(matrix: list[list[int]]) -> list[list[int]]:
            n = len(matrix)
            return [[matrix[n - 1 - j][i] for j in range(n)] for i in range(n)]

        for _ in range(4):
            if mat == target:
                return True
            mat = rotate(mat)
        return False

File: 03-march/22-determine-whether-matrix-can-be-obtained-by-rotation/solution.java
class Solution {
    public boolean findRotation(int[][] mat, int[][] target) {
        int n = mat.length;

        for (int rotation = 0; rotation < 4; rotation++) {
            if (equal(mat, target)) {
                return true;
            }
            mat = rotate(mat, n);
        }
        return false;
    }

    private boolean equal(int[][] a, int[][] b) {
        int n = a.length;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (a[i][j] != b[i][j]) {
                    return false;
                }
            }
        }
        return true;
    }

    private int[][] rotate(int[][] matrix, int n) {
        int[][] res = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                res[i][j] = matrix[n - 1 - j][i];
            }
        }
        return res;
    }
}
